### PR TITLE
Silicon Sages - Epic 7.1 - Multiple Choice Assessments Generator

### PR DESCRIPTION
**Pull request by AaronSosaRamos**
# Pull Request - Epic 7.1 - Multiple Choice Assessments Generator
By: Silicon Sages

## Summary
This PR allows educators to generate multiple choice assessments based in the desired topic, number of questions and language. Moreover, educators can specify the context with the extended file support implemented with the following file types for a better domain-specific adaptation:

- PDF
- CSV
- TXT
- MD
- URL (Such as Wikipedia or other source with info.)
- PPTX
- DOCX
- XLS / XLSX
- XML
- GOOGLE DOCS
- GOOGLE SHEETS
- GOOGLE SLIDES
- GOOGLE PDF
- IMAGES (PNG / JPG / JPEG)
- MP3 

## Changes
- Reused the document loaders from Epic 7.2 #61.
- Updated the existent `tool.py` because we integrated the new document loaders and innovative approaches that LangChain v. 0.2 enhances for optimal AI pipelines.
- Updated the existent `core.py` for decopling the request args. as well as implementing the new workflow from tool.py appropriately.
- Reused the util file `allowed_file_extensions_dynamo.py` for a robust approach in the file type management from Epic 7.2 #61.
- Decoupling Quizzify's args. as a best practice for allowing flexible maintainability similarly as Epic 7.3 #80.
- Created new Error Classes for improving the error mapping process.
- Enabled a Computer Vision approach for generating multiple choice assessments from images as an innovative proposal for the project.
- Thanks to @asleem for developing the additional MP3 document loader for processing audio with 2 innovative methods. Additionally, she has given support for solving the bugs related to the files which open in preview.
- Thanks to @lancegosu for his extensive research and contribution with CoT prompting for structured data.
- Thanks to @ashadiiii for her continuous support in the innovative strategies for the multimodal support 

## Testing
Assumption: Latency has not increased since the AI pipeline uses the built-in document loaders from #81. 
image

New document loaders work for MP3 files:
image

## Results
All document loaders are working appropriately with the optimal results.
image

## Notes

For Google Docs, Slides, Sheets and PDF files: 
- Those files need to be shared as **public** in Google Drive.
- They need to be uploaded to Google Drive. If a Google File is created in Google Drive, it needs to be downloaded and then uploaded to be detected. This was the most appropriate approach that we discovered as the GoogleDriver loader from LangChain works with OAuth2, which is not appropriate for production deployment.

## How to Test
1. Clone the repo in your local

2. Create and activate virtual environment

3. Use pip install -r requirements.txt to install required libraries.

4. Create .env file with ENV_TYPE, GCP_PROJECT_ID and GOOGLE_API_KEY fields. Env type is dev, gcp_project_id is your project id from cloud console project, google_api_key is your api key from AI studio.

5. Then type ./local-start.sh to start the application.

6. Add sample requests and responses for each file type as mentioned above in screenshots and test it.

### Schema:
```python
class QuizzifyArgs(BaseModel):
    topic: str
    n_questions: int
    file_url: str
    file_type: str
    lang: Optional[str] = "en"
```

### Example:
```json
{
   "user":{
      "id":"string",
      "fullName":"string",
      "email":"string"
   },
   "type":"chat",
   "tool_data":{
      "tool_id":0,
      "inputs":[
         {
            "name":"quizzify_args",
            "value":{
               "topic":"Machine Learning",
               "n_questions":3,
               "file_url":"https://www.youtube.com/watch?v=W01tIRP_Rqs",
               "file_type":"youtube_url",
               "lang":"en"
            }
         }
      ]
   }
}
```

### Schema:
```python
class QuestionChoice(BaseModel):
    key: str = Field(description="A unique identifier for the choice using letters A, B, C, or D.")
    value: str = Field(description="The text content of the choice")
class QuizQuestion(BaseModel):
    question: str = Field(description="The question text")
    choices: List[QuestionChoice] = Field(description="A list of choices for the question, each with a key and a value")
    answer: str = Field(description="The key of the correct answer from the choices list")
    explanation: str = Field(description="An explanation of why the answer is correct")

    model_config = {
        "json_schema_extra": {
            "examples": """ 
                {
                "question": "What is the capital of France?",
                "choices": [
                    {"key": "A", "value": "Berlin"},
                    {"key": "B", "value": "Madrid"},
                    {"key": "C", "value": "Paris"},
                    {"key": "D", "value": "Rome"}
                ],
                "answer": "C",
                "explanation": "Paris is the capital of France."
              }
          """
        }

      }
```

### Example: 
```json 
{
  "data": [
    {
      "question": "Which of the following is an example of a regression algorithm?",
      "choices": [
        {
          "key": "A",
          "value": "Support vector machines"
        },
        {
          "key": "B",
          "value": "Decision trees"
        },
        {
          "key": "C",
          "value": "Linear regression"
        },
        {
          "key": "D",
          "value": "Random forests"
        }
      ],
      "answer": "C",
      "explanation": "Linear regression is an example of a regression algorithm, which outputs a continuous value such as price or probability."
    },
    {
      "question": "Which type of machine learning uses labeled input?",
      "choices": [
        {
          "key": "A",
          "value": "Supervised learning"
        },
        {
          "key": "B",
          "value": "Unsupervised learning"
        },
        {
          "key": "C",
          "value": "Reinforcement learning"
        },
        {
          "key": "D",
          "value": "Transfer learning"
        }
      ],
      "answer": "A",
      "explanation": "Supervised learning is a type of machine learning that uses labeled input, meaning that the data is already categorized or labeled. This allows the model to learn the relationship between the input and output data."
    },
    {
      "question": "Which of the following is NOT an example of a classification algorithm?",
      "choices": [
        {
          "key": "A",
          "value": "Support vector machines"
        },
        {
          "key": "B",
          "value": "Decision trees"
        },
        {
          "key": "C",
          "value": "Linear regression"
        },
        {
          "key": "D",
          "value": "Random forests"
        }
      ],
      "answer": "C",
      "explanation": "Linear regression is an example of a regression algorithm, not a classification algorithm."
    }
  ]
}
```

### CODE REVIEW

File: README.md
Comment (hash2004) [inline README.md:1]: There is no need to change the README.md file. You can check the latest from the STAGING branch.

File: app/features/quizzify/document_loaders.py [hunk 1]
```diff
@@ -0,0 +1,376 @@
+from langchain_community.document_loaders import YoutubeLoader
+from langchain_community.document_loaders import PyPDFLoader
+from langchain_community.document_loaders.csv_loader import CSVLoader
```
Comment (hash2004) [inline app/features/quizzify/document_loaders.py:3]: You can import all the document loaders in one line for best practice, like: 
from langchain_community.document_loaders import TextLoader, UnstructuredURLLoader, UnstructuredPowerPointLoader, Docx2txtLoader, UnstructuredExcelLoader, UnstructuredXMLLoader

File: requirements.txt
Comment (hash2004) [inline requirements.txt:1]: Add fastapi[standard] or use the requirements.txt file from the STAGING branch.

**Review by hash2004 — CHANGES_REQUESTED** [review]
Please see requested changes. Good work though!

File: README.md
Comment (AaronSosaRamos) [reply README.md:1]: Done, I updated the README.md file

File: requirements.txt
Comment (AaronSosaRamos) [reply requirements.txt:1]: Done!

File: app/features/quizzify/document_loaders.py (hunk 1 above)
Comment (AaronSosaRamos) [reply app/features/quizzify/document_loaders.py:3]: Done!

**Review by huzaifakhan04 — CHANGES_REQUESTED** [review]
This feature functions properly for links to files that automatically download. However, it **_does not seem to work for files that open in preview mode_**. For example, the following request fails:

**Request:**
```json
{
   "user":{
      "id":"string",
      "fullName":"string",
      "email":"string"
   },
   "type":"chat",
   "tool_data":{
      "tool_id":0,
      "inputs":[
         {
            "name":"quizzify_args",
            "value":{
               "topic":"Machine Learning",
               "n_questions":3,
               "file_url":"https://sebastianraschka.com/pdf/lecture-notes/stat479fs18/01_ml-overview_notes.pdf",
               "file_type":"pdf",
               "lang":"en"
            }
         }
      ]
   }
}
```

**Response:**
```json
{
  "status": 500,
  "message": "Error in executor: Unsupported file type"
}
```

From a user's perspective, they often provide links that don’t trigger automatic downloads but instead open in preview. There should be a way to handle these types of links as well.

The same issue occurs with other file types. Here are the tested formats:

- YouTube Link (**Bug:** ❌)
- `.pdf` File Link (**Bug:** ✅)
- `.csv` File Link (**Bug:** ✅)
- `.txt` File Link (**Bug:** ✅)
- `.pptx` File Link (**Bug:** ✅)

Please address this as soon as possible so we can merge your PR. Thank you! :)

**Pull request merged**